### ISSUE TITLE
Doesnt retain environment variables from bash (in WSL)

### ISSUE THREAD

**Issue by Aetherbase** [arkane-systems/genie#35]
When we start a new shell , the environment variables (especially Path) are modified when run in wsl.This should not happen . It crashes many environment setups.

**Comment by cerebrate** [comment]
This is a consequence of running inside the genie bottle; _systemd(1)_ runs as pid 1 init, and as such expects to set up the system environment itself, and for PAM to set up user environments when a user session is created (which, when using _genie -s_ or _genie -c_, is done via _runuser(1)_). This isn't a design choice of mine; it's just how containers/systemd/PAM work in Linux.

Genie actually incorporates a mechanism for cloning some outside-bottle environment variables already to ensure that interop works from inside the bottle; you can see this in _/lib/genie/dumpwslenv.sh_ (which saves the WSL-specific environment variables to a temp file under _/run_) and  _/lib/systemd/system-environment-generators/10-genie-envar.sh_ (which sets them in the bottle environment).

It would be possible to add additional variables to _dumpwslenv.sh_, or better yet, to develop your own mechanism along these lines (to prevent future genie releases from overwriting it); this, though, should be done with care, since depending on what you have installed inside the bottle, blanket overwriting of the systemd/PAM environment will break things: for example, on my system, simply overwriting the PATH would remove several directories added by systemd services, and replacing the environment as a whole would break yet more: dbus, homebrew, snapcraft, xdg, etc.

**Comment by arlllk** [comment]
I just thought of this, but, it it could be possible to use a .conf in /etc to setup there the environment variables that you wish to pass on, it's much easier than changing those shell scripts, the program could read the config file and do a foreach with all variables, if there is a variable that is written in the .conf file that don't exist in the environment, it could show like a warning at the time of entering to genie environment,

**Comment by cerebrate** [comment]
I like the idea, but am unlikely to have time to implement it myself any time soon; I've got a hard deadline coming up on another project that's eating most of my attention. (Pull requests gratefully accepted, on the other hand.)

PATH in particular is still likely to need special handling, though: that in particular probably needs a merge rather than a replace to ensure nothing breaks on either side. Not that it wouldn't be a good step forward anyway.

**Comment by BrainStone** [comment]
Generally speaking environment variables should be preserved. Some may need special handling but other than that they should migrated.  
Also the cwd should be preserved.

**Comment by cerebrate** [comment]
See here for the current status of things:

https://github.com/arkane-systems/genie/wiki/A-note-on-environment-variables-and-cwd

tl;dr blind copying of environment variables or $PATH violates login semantics for _genie -s_ and is likely to break things under _genie -c_ . cwd is preserved when appropriate; i.e. -c only.

**Comment by BrainStone** [comment]
I mean it's fine that it doesn't preserve the path by default. But there should be a flag for it it allow it preserve the path.

The main use case for this is to make the "Open linux shell here" context option work again if you added the using the bottle in the .profile file.

**Comment by cerebrate** [comment]
Now that I think of it, that context menu option would make much more sense in the genie context if it opened a _non-_login shell, semantically speaking. So the best option here may be to edit the invocation that that option uses to invoke `genie -c zsh` (substitute own preferred shell here); i.e., edit the _HKCR\Directory\shell\WSL\command_ default key to be

```
wsl.exe --cd "%V" genie -c "sh -c `getent passwd $LOGNAME | cut -d: -f7`"
```
...which should preserve cwd and $PATH (including the Windows path add-ons), although it does not recreate the systemd path adds. Seems to work in my quick tests here, anyway, and is probably less klugey than trying to do path merging inside genie. If it shows no problems after further poking, I'll write it up for the wiki.

**Comment by cerebrate** [comment]
(And done. See here: https://github.com/arkane-systems/genie/wiki/%22Open-a-linux-shell%22-and-genie )

**Comment by BrainStone** [comment]
Awesome! Thank you very much.
